### ISSUE TITLE
scruff

### ISSUE THREAD

**Issue by zhewa** [Bioconductor/Contributions#756]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/compbiomed/scruff

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
@nturaga will review your package. In your vignette, do _not_ write to a local disk location `./...` but instead used `tempfile()`.

**Comment by zhewa** [comment]
Hi reviewer,

It's been more than an hour and there's still no package builder report yet. Should I issue another version bump?

Thanks

**Comment by zhewa** [comment]
Hi @nturaga ,

There is this persistent error "subscript is a NSBS object that is incompatible with the current subsetting operation" on the build systems but I can not reproduce this error on my local machine with Linux environment. Can you check what is going on here?

Thank you,
Zhe

**Comment by zhewa** [comment]
Hi @nturaga 

Oddly the error I mentioned above disappeared in the latest commit though I did not change anything except bumping the version. Now there are only two errors remaining:

1. ERROR: Maintainer must subscribe to the bioc-devel mailing list.
      Subscribe here: https://stat.ethz.ch/mailman/listinfo/bioc-devel
2. * checking package dependencies ... ERROR
Package suggested but not available: 'Rsubread'

The suggested packages are required for a complete check.
Checking can be attempted without them by setting the environment
variable _R_CHECK_FORCE_SUGGESTS_ to a false value.

For the first error, I already registered on the bioc-devel mailing list so I'm not sure how I can fix this.
For the second error, because some of the functions in scruff package relies on Rsubread package which is not available on Windows environment, so I put Rsubread package requirement to suggest instead of import. Do you want it removed from the suggest list?

Thank you,
Zhe

**Comment by nturaga** [comment]
Hi @zhewa ,

`Rsubread` is not supported on windows, so you will get that error. @mtmorgan might have a suggestion for you about how to avoid building your package on windows. I think it's easy to set a `.BBSoptions` file in your package. It should look like this. 

```
cat .BBSoptions                                                                                                                                                                                                        master
# Default value for 'UnsupportedPlatforms' is the empty string.
# Examples:
#   - for a package that is supported on all platforms except on Windows:
#       UnsupportedPlatforms: win
#   - for a package that is supported on all platforms execept on Mac OS X and
#     Solaris:
#       UnsupportedPlatforms: mac, churchill
UnsupportedPlatforms: win
```


# Package Review

## R CMD build

ok

##  R CMD INSTALL

ok

## Build Report

* The build report is not as clean as it could be. There are ERROR and
  NOTE messages which can be resolved.
  
  
		* Checking unit tests...
			* NOTE: Consider adding unit tests. We strongly encourage them. See
		http://bioconductor.org/developers/how-to/unitTesting-guidelines/.
		* Checking for <<-...
			Found <<- in R/scruffFunctions.R (line 107, column 15)
			* NOTE: Avoid '<<-' if possible (found in 1 files)
		* Checking function lengths...............
			The longest function is 294 lines long
			The longest 5 functions are:
			.demultiplexUnit() (R/demultiplex.R, line 201): 294 lines
			gview() (R/scruffFunctions.R, line 206): 203 lines
			countUMI() (R/countUMI.R, line 59): 185 lines
			demultiplex() (R/demultiplex.R, line 49): 149 lines
			alignRsubread() (R/alignRsubread.R, line 55): 147 lines
		* Checking exported objects have runnable examples...
			* NOTE: Consider adding runnable examples to the following man
			  pages which document exported objects:
			  alignRsubread.Rd
		* Checking formatting of DESCRIPTION, NAMESPACE, man pages, R source,
		  and vignette source...
			* NOTE: Consider shorter lines; 207 lines (7%) are > 80 characters
			  long.
			First 6 lines:
			  R/alignRsubread.R:3 #' This function is \strong{not} available in Windo...
			  R/alignRsubread.R:5 #' @param sce A \code{SingleCellExperiment} object ...
			  R/alignRsubread.R:6 #' @param index Path to the \code{Rsubread} index o...
			  R/alignRsubread.R:7 #' @param unique Argument passed to \code{align} fu...
			  R/alignRsubread.R:8 #' @param nBestLocations Argument passed to \code{a...
			  R/alignRsubread.R:9 #' @param format File format of sequence alignment ...
			* NOTE: Consider multiples of 4 spaces for line indents, 1005
			  lines(35%) are not.
			First 6 lines:
			  R/alignRsubread.R:56                           index,
			  R/alignRsubread.R:57                           unique = FALSE,
			  R/alignRsubread.R:58                           nBestLocations = 1,
			  R/alignRsubread.R:59                           format = "BAM",
			  R/alignRsubread.R:60                           outDir = "./Alignment",
			  R/alignRsubread.R:61                           cores = max(1, parallel:...
			  
** See http://bioconductor.org/developers/how-to/coding-style/ ** 

* Reindent your code according to the Bioconductor style guide.

## DESCRIPTION

* Please remove `Author` and `Maintainer` if you are using `Authors@R`. 

## NAMESPACE

ok

## man

ok

## vignette

* Do not include the PDF file for the vignette, just the `Rmd`
  file. The PDF will be built from the `Rmd`.

* There is an empty graphics canvas being shown when the vignette is
  being run.

* Line wrap comments by breaking the lines easier for reading.

* Explain each plot in your vignette and try to make the vignette as
  detailed as possible for a novice user. Point to the appropriate
  resources when using other packages like SingleCellExperiment.

## R

* Good to see single cell infrastructure like SingleCellExperiment
  being reused.

* No need to use `message()` and `paste()` together to concatenate. Just
  use `message()`. Take a look at all these `message()` calls. You don't
  need it unless you are trying to recyle vectors within the message
  statement.
  
  Similarly, with `stop()`, you don't need to use `paste()`.

```
R/alignRsubread.R
78:  message(paste(Sys.time(), "Start alignment ..."))
97:    message(paste(Sys.time(), "... Delete (if any) existing alignment results"))
115:  message(paste(Sys.time(), "... Creating output directory", outDir))
173:  message(paste(Sys.time(), paste(
195:  message(paste(Sys.time(), "... Add alignment information to SCE colData."))
199:  message(paste(Sys.time(), "... Alignment done!"))

R/countUMI.R
69:  message(paste(Sys.time(), "Start UMI counting ..."))
98:  message(paste(Sys.time(),
105:  message(paste(Sys.time(),
144:  message(paste(Sys.time(), paste(
158:  message(paste(Sys.time(),
227:  message(paste(Sys.time(), paste(
241:  message(paste(Sys.time(), "... UMI counting done!"))

R/demultiplex.R
70:  message(paste(Sys.time(), "Start demultiplexing ..."))
74:    message(paste(Sys.time(), "All files in", outDir,  "will be deleted ..."))
158:  message(paste(
195:  message(paste(Sys.time(), "... Demultiplex done!"))

R/scruff.R
113:  message(paste(Sys.time(), "Start running scruff ..."))
166:  message(paste(Sys.time(), "Finished running scruff ..."))

R/scruffFunctions.R
18:    message(paste(..., sep = sep))
61:    message(paste(Sys.time(), "... TxDb file", gtf.db.file, "does not exist"))
62:    message(paste(Sys.time(), "... Creating TxDb object", gtf.db.file))
```

* Use Bioconductor's core infrastructure, `BiocParallel` instead of
  parallelizing with
  foreach. https://bioconductor.org/packages/release/bioc/html/BiocParallel.html.
  
  You can use the BiocParallelParam of your choice, and pass it to a
  `bplapply(X, FUN, ... ,BPPARAM)` method instead. The vignette holds very detailed
  explainations of how to use the package. Figuring out how many cores
  to use and such bookkeeping is already done behind the scenes by
  BiocParallel. 
  
  `BiocParallel::BatchtoolsParam()` is particularly suited for
  parallelizing on a cluster.
  
```
R/alignRsubread.R
61:                          cores = max(1, parallel::detectCores() / 2),
118:  # parallelization
120:    parallel::makeCluster(cores, outfile = logfile)
122:    parallel::makeCluster(cores)
169:  parallel::stopCluster(cl)

R/countUMI.R
64:                     cores = max(1, parallel::detectCores() / 2),
109:  # parallelization
111:    parallel::makeCluster(cores, outfile = logfile)
113:    parallel::makeCluster(cores)
139:  parallel::stopCluster(cl)

R/demultiplex.R
66:                        cores = max(1, parallel::detectCores() / 2),
100:  # parallelization
102:    parallel::makeCluster(cores, outfile = logfile)
104:    parallel::makeCluster(cores)
156:  parallel::stopCluster(cl)

R/scruff.R
108:                   cores = max(1, parallel::detectCores() / 2),
```

* Declare the non-exported functions in `scruff_functions.R` outside
  the scope of just, `gview`. Makes reading the code and following it
  easier. It is also a good practice so that you can document what
  each function is doing seperately.


Please make these corrections and leave a detailed reply. Once this is done I will review your package once again.

**Comment by zhewa** [comment]
Hi @nturaga,

- We added the `.BBSoptions` file as suggested.

- Build Report
  - We added a few unit tests.
  - We use `<<-` in `.bcCorrectMem` function in `scruffFunctions.R` file to assign values to a list element in parent environment. `assign()` does not work well with list element assignment so we sticked with `<<-`.
  - Some functions have long lengths due to historical reasons and partly due to that these functions have their own specific individualized roles. We plan to modularize them in future updates.
  - We did not add an runnable example for `alignRsubread` function because it will take more than 5 minutes for the R CMD build check which results in a warning.
  - We reindented the codes.

- DESCRIPTION
  - `Author` and `Maintainer` are removed.

- vignette
  - The PDF file is removed.
  - An empty first page is shown when the `qcplots` function is called. This is probably due to the incompatibility of `gridExtra::marrangeGrob` function and `ggplot` objects. Currently we are not able to fix this issue. 
  - We reindented the codes in vignette.
  - We added instructions on how to read the plots.

- R
  - We removed `paste()` calls within `message()` and `stop()` function calls.
  - We switched to `BiocParallel` for parallelization.
  - We moved `rview` and `gview` functions to separate files.

**Comment by nturaga** [comment]
Please make sure your build report is without ERRORs and as few NOTEs as possible 

```
* Checking for bioc-devel mailing list subscription...
    * ERROR: Maintainer must subscribe to the bioc-devel mailing list.
      Subscribe here: https://stat.ethz.ch/mailman/listinfo/bioc-devel
* Checking for support site registration...
    * ERROR: Maintainer must register at the support site; visit
      https://support.bioconductor.org/accounts/signup/ .
```

**Comment by zhewa** [comment]
Hi @nturaga 

We found out that these two errors are due to the typo in the email address in the DESCRIPTION file. This is a fault on our side. Now it is fixed. 

Thank you so much for your time in reviewing our package.

Zhe

**Comment by zhewa** [comment]
Hi @nturaga 

We fixed the errors. Please let us know if any other modifications are required.

Thank you,
Zhe

**Comment by nturaga** [comment]
Hi @zhewa 

I'll review your package again shortly and let you know.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/zhewa.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"scruff\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/scruff

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
